File: packages/modules/api/rust/src/api/telegram.org/telegram.rs
use anyhow::Result;
use reqwest::Client;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug)]
pub struct TelegramResponse {
    pub ok: bool,
    pub result: serde_json::Value,
}

pub struct TelegramClient {
    client: Client,
    bot_token: String,
    pub base_url: String,
}

impl TelegramClient {
    pub fn new(bot_token: String, base_url: Option<String>) -> Self {
        Self {
            client: Client::new(),
            bot_token,
            base_url: base_url.unwrap_or_else(|| "https://api.telegram.org".to_string()),
        }
    }

    pub async fn send_message(&self, chat_id: &str, text: &str) -> Result<TelegramResponse> {
        let url = format!("{}/bot{}/sendMessage", self.base_url, self.bot_token);
        let resp = self
            .client
            .post(&url)
            .json(&serde_json::json!({"chat_id": chat_id, "text": text}))
            .send()
            .await?
            .json::<TelegramResponse>()
            .await?;
        Ok(resp)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use wiremock::matchers::{method, path};
    use wiremock::{Mock, MockServer, ResponseTemplate};

    #[tokio::test]
    async fn test_send_message() {
        let mock_server = MockServer::start().await;

        Mock::given(method("POST"))
            .and(path("/botTEST_TOKEN/sendMessage"))
            .respond_with(ResponseTemplate::new(200).set_body_json(serde_json::json!({
                "ok": true,
                "result": {"message_id": 1}
            })))
            .mount(&mock_server)
            .await;

        let client = TelegramClient::new("TEST_TOKEN".to_string(), Some(mock_server.uri()));
        let result = client.send_message("12345", "Hello").await.unwrap();
        assert!(result.ok);
    }
}

File: packages/modules/api/python/src/api/telegram.org/telegram.py
from enum import StrEnum

import httpx

BASE_URL = "https://api.telegram.org/bot"
INVALID_TOKEN = "Invalid token"


class ParseMode(StrEnum):
    HTML = "html"
    MARKDOWN = "markdown"


class TelegramClient:
    def __init__(self, token: str) -> None:
        self._token = token
        self._client = httpx.AsyncClient()

    async def _post(self, method: str, json_body: dict | None = None) -> dict:
        url = f"{BASE_URL}{self._token}/{method}"
        response = await self._client.post(url, json=json_body)
        return response.json()

    async def send_message(
        self,
        chat_id: int,
        message: str,
        parse_mode: ParseMode = ParseMode.MARKDOWN,
    ) -> None:
        if not self._token:
            raise ValueError(INVALID_TOKEN)
        if not chat_id:
            raise ValueError("Invalid chatId")
        if not message:
            raise ValueError("Invalid message")
        params = {
            "chat_id": chat_id,
            "text": message,
            "parse_mode": parse_mode.value,
        }
        url = f"{BASE_URL}{self._token}/sendMessage"
        await self._client.post(url, params=params)

    async def set_webhook(self, url: str) -> dict:
        if not self._token:
            raise ValueError(INVALID_TOKEN)
        if not url:
            raise ValueError("Invalid url")
        return await self._post("setWebhook", {"url": url})

    async def delete_webhook(self, url: str) -> dict:
        if not self._token:
            raise ValueError(INVALID_TOKEN)
        if not url:
            raise ValueError("Invalid url")
        return await self._post("deleteWebhook", {"url": url})

    async def get_webhook_info(self) -> dict:
        if not self._token:
            raise ValueError(INVALID_TOKEN)
        return await self._post("getWebhookInfo")
